### DISCUSSION TITLE
Setting DataGrid selected items from code

### DISCUSSION THREAD

**Discussion by wgraaf** [AvaloniaUI/Avalonia#11425] [Q&A] [answered]
How can a selection of multiple DataGrid rows be set from code?

DataGrid.SelectedItems does not appear to be bindable and it is also not settable.

In my scenario, I have a tab control with items  containing a datagrid:

      <TabControl.ContentTemplate>
        <DataTemplate>
          <DataGrid Name="DataGrid"
           .....
          </DataGrid>
        </DataTemplate>
      </TabControl.ContentTemplate>

Switching tabs clears any selection made in the datagrid; I would like to restore the selection upon switching back to a tab.

**Comment by wgraaf** [comment]
In general: how can a multiple row selection state of a DataGrid be set and/or restored?

**Comment by maxkatz6** [comment, marked as answer]
DataGrid has SelectedItems collection property which you can mutate (add/remove items). And SelectionChanged event to react on these changes.

**Reply by wgraaf** [reply]
Thanks, that works.
Too bad that SelectedItems is not bindable, though.
Is there an alternative to exposing it to the view model other than using code behind?

**Reply by maxkatz6** [reply]
As I said above, it has SelectionChanged event, so you can manually bind it.

**Reply by timunie** [reply]
> Is there an alternative to exposing it to the view model other than using code behind?

I think a custom Behavior (https://github.com/AvaloniaUI/Avalonia.Xaml.Behaviors) or an attached property can work. In these you can do whatever is needed to sync both collections. There are some samples for WPF out there in the www.

**Reply by wgraaf** [reply]
Using this to subcribe to SelectionChanged event:

>      <DataGrid>
> 
>             <Interaction.Behaviors>
>               <EventTriggerBehavior EventName="SelectionChanged">
>                 <InvokeCommandAction Command="{Binding SelectionChangedCommand}"></InvokeCommandAction>
>               </EventTriggerBehavior>
>             </Interaction.Behaviors>
>             
>             ...
>      </DataGrid>
> 

And a Reactive command in the view model:

>       SelectionChangedCommand = ReactiveCommand.Create(SelectionChanged);

With:

    

>       public void SelectionChanged() {...}

**Reply by shimpe** [reply]
Maybe useful to add that one must 

`dotnet add package Avalonia.Xaml.Behaviors`

to avoid getting cryptic errors. 

And you can use

` <InvokeCommandAction Command="{Binding SelectionChangedCommand}" PassEventArgsToCommand="true"/>`

to ensure that the SelectionChanged handler receives an argument SelectionChangedEventArgs args.

**Comment by wgraaf** [comment]
Restoring the selection on a tab switch works, but the scroll position is still lost.

Do you have any suggestions on capturing and restoring that?

The (vertical) ScrollBar component has a Value property that might be used for this, if it can be accessed.

**Reply by timunie** [reply]
You can call `myDataGrid.ScrollIntoView` I guess. But it needs to be called on Dispatcher with priority Loaded iirc.

**Reply by wgraaf** [reply]
How do I capture the bottom visible row, so I can restore the scroll position upon a tab switch?
There is a LoadingRow event that might help out, but I don't seem to be able to get a parameter passed into the event handler
(see issue https://github.com/AvaloniaUI/Avalonia/discussions/11505)

**Reply by wgraaf** [reply]
I found a way to capture the item for the ScrollIntoView call , avoiding the event handler argument.
It is ugly:

    private int _verticalScrollIndex = 0;
    private void VerticalScroll(ScrollEventArgs e)//DEBUGx args always null??
    {
      //capture index of last visible row
      var sb = MainWindowViewModel.Instance.VerticalScrollBar;
      if (sb == null)
      {
        return;
      }

      var rowHeight = 24;//get from component
      var startIdx = (int)Math.Ceiling(sb.Value / rowHeight);
      var offset = (int)Math.Floor(sb.ViewportSize / rowHeight) - 1;
      var endIdx = startIdx + offset;

      if (sb.Value == sb.Maximum - 1)//occurs when scrolling to the end, then reversing scroll direction
      {
        --startIdx;
      }
      _verticalScrollIndex = startIdx;
    }

However, as alluded to, calling DataGrid.ScrollIntoView() from the TabControl SelectedItem setter does not do the trick.
Calling it from a menu command does work.

Dispatching like below does not work:

      Dispatcher.UIThread.InvokeAsync((Action)delegate ()
      {
        _dataGrid.ScrollIntoView(item, null);
      });

What do you mean by 'call on Dispatcher with priority Loaded iirc' ?

**Reply by wgraaf** [reply]
Dispatching with DispatcherPriority.ApplicationIdle or DispatcherPriority.ContextIdle appears to work!

**Reply by wgraaf** [reply]
Dispatcher.UIThread.InvokeAsync((Action)(() => _dataGrid.ScrollIntoView(item, null)), DispatcherPriority.ContextIdle);
